### PR TITLE
[terra-application-navigation] Adding custom event to trigger popup and drawer menu closure.

### LINKED ISSUE #954: [Terra Application Navigation] Add compact drawer open/close functionality
# Feature Request

## Description
I am currently consuming the Application Navigation component and there are scenarios where I need to know if the compact drawer is currently open or not and also need the ability to close it if need be.

In the mobile version of my application, if a user opens up the drawer and backgrounds the app, when the app is brought back into the foreground the drawer will still be open. This is problematic in the event that a user receives a notification and selects that notification. When this happens, my application needs to navigate through the app and get the user to a specific view. Although the navigational changes are correctly made, they are all done in the "background" because the drawer is still open upon foregrounding.

## @ Mentions
@ryanthemanuel 
@tbiethman 
@dkasper-was-taken

### PR DESCRIPTION
**Pull request by BenBoersma**
### Summary

Adding custom event to trigger popup and drawer menu closure. Custom event is `terra-application-navigation.dismiss-menu`.

Closes #954

### Deployment Link

https://terra-framework-deployed-pr-#.herokuapp.com/

### Testing

Added new WDIO tests that dispatch the custom event while drawer and popup are open to close.

### Additional Details

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js [hunk 1]
```diff
@@ -122,6 +122,26 @@ Terra.describeViewports('ApplicationNavigation - Large', ['large'], () => {
...
     Terra.it.validatesElement();
   });
+
+  describe('Should close open popup when custom event is dispatched', () => {
+    before(() => browser.url('/#/raw/tests/terra-application-navigation/application-navigation/application-navigation'));
+
+    it('launch popup', () => {
+      browser.click('[data-application-header-utility="true"]');
+      browser.waitForVisible('[data-terra-popup-content="true"]');
+      browser.pause(50);
```
Comment (emilyrohrbough) [inline packages/terra-application-navigation/tests/wdio/application-navigation-spec.js:132]: why is the pause needed?

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js [hunk 2]
```diff
@@ -122,6 +122,26 @@ Terra.describeViewports('ApplicationNavigation - Large', ['large'], () => {
...
+    });
+
+    Terra.it.validatesElement('1. Popup Open');
+
+    it('close popup', () => {
+      browser.execute(() => {
+        window.dispatchEvent(new Event('terra-application-navigation.dismiss-menu'));
+      })
```
Comment (emilyrohrbough) [inline packages/terra-application-navigation/tests/wdio/application-navigation-spec.js:140]: We need to use func syntax tor un correctly in IE: https://github.com/cerner/terra-framework/pull/1109/files#diff-06ad68eca863691057a7cdb60e83ececR522-R523

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js [hunk 3]
```diff
@@ -296,6 +316,26 @@ Terra.describeViewports('ApplicationNavigation - Small', ['small'], () => {
...
     Terra.it.validatesElement({ selector: '#root' });
   });
+
+  describe('ApplicationNavigation should close open drawer when custom event is dispatched', () => {
+    before(() => browser.url('/#/raw/tests/terra-application-navigation/application-navigation/application-navigation'));
+
+    it('open drawer', () => {
+      browser.waitForVisible('[data-compact-header-toggle="true"]');
+      browser.moveToObject('[data-compact-header-toggle="true"]').leftClick();
+      browser.pause(250);
```
Comment (emilyrohrbough) [inline packages/terra-application-navigation/tests/wdio/application-navigation-spec.js:326]:
```suggestion
```

**Comment by emilyrohrbough** [comment]
@BenBoersma Looking at the last update on the Issue this closed, there was no mention of adding an event listener to handle this. Did something change?

**Comment by BenBoersma** [comment]
> @BenBoersma Looking at the last update on the Issue this closed, there was no mention of adding an event listener to handle this. Did something change?

Yes. Open actually testing the changes that were listed in the issue for the spike we discovered some discrepency in out it worked between IOS and Android. The original changes would work in one and not the other. After some discussion we moved to the current format of providing a custom event.

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js (hunk 1 above)
Comment (BenBoersma) [reply packages/terra-application-navigation/tests/wdio/application-navigation-spec.js]: I don't believe it is, was leftover when I copied some code from other tests. Fixed here https://github.com/cerner/terra-framework/pull/1255/commits/d1acd6f55250e3e76cca9dae3979ca566a864db9#diff-85a7fe8bb38bbc87cbbac08c28ef3370L132.

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js (hunk 3 above)
Comment (BenBoersma) [reply packages/terra-application-navigation/tests/wdio/application-navigation-spec.js]: Removed https://github.com/cerner/terra-framework/pull/1255/commits/d1acd6f55250e3e76cca9dae3979ca566a864db9#diff-85a7fe8bb38bbc87cbbac08c28ef3370L326

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js (hunk 2 above)
Comment (BenBoersma) [reply packages/terra-application-navigation/tests/wdio/application-navigation-spec.js]: Fixed https://github.com/cerner/terra-framework/pull/1255/commits/d1acd6f55250e3e76cca9dae3979ca566a864db9#diff-85a7fe8bb38bbc87cbbac08c28ef3370R138

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js [hunk 4]
```diff
@@ -122,6 +122,26 @@ Terra.describeViewports('ApplicationNavigation - Large', ['large'], () => {
...
+      browser.click('[data-application-header-utility="true"]');
+      browser.waitForVisible('[data-terra-popup-content="true"]');
+    });
+
+    Terra.it.validatesElement('1. Popup Open');
+
+    it('close popup', () => {
+      // eslint-disable-next-line prefer-arrow-callback
+      browser.execute(function dispatch() {
+        window.dispatchEvent(new Event('terra-application-navigation.dismiss-menu'));
```
Comment (yuderekyu) [inline packages/terra-application-navigation/tests/wdio/application-navigation-spec.js:139]: ie doesn't support event constructors, so we'd need to do something similar to this 
https://github.com/cerner/terra-application/blob/main/tests/wdio/application-base/private/test-overrides-spec.js#L6-L9

File: packages/terra-application-navigation/src/ApplicationNavigation.jsx [hunk 5]
```diff
@@ -321,6 +321,19 @@ const ApplicationNavigation = ({
...
+  useEffect(() => {
+    const forceCloseMenu = () => {
+      setDrawerMenuIsOpen(false);
+      setPopupMenuIsOpen(false);
+    };
+
+    window.addEventListener('terra-application-navigation.dismiss-menu', forceCloseMenu);
+
+    return () => {
+      window.removeEventListener('terra-application-navigation.dismiss-menu', forceCloseMenu);
```
Comment (yuderekyu) [inline packages/terra-application-navigation/src/ApplicationNavigation.jsx:333]: nitpick - we can save the event name as a constant then use that to setup and tear down the event listener

**Review by tbiethman — APPROVED** [review]

**Review by mjhenkes — APPROVED** [review]

File: packages/terra-application-navigation/tests/wdio/application-navigation-spec.js (hunk 4 above)
Comment (BenBoersma) [reply packages/terra-application-navigation/tests/wdio/application-navigation-spec.js]: Fixed here https://github.com/cerner/terra-framework/pull/1255/commits/60dc2ba504e31fd36dbc694fc4b95e9d046848f2#diff-85a7fe8bb38bbc87cbbac08c28ef3370R139

File: packages/terra-application-navigation/src/ApplicationNavigation.jsx (hunk 5 above)
Comment (BenBoersma) [reply packages/terra-application-navigation/src/ApplicationNavigation.jsx]: renamed here https://github.com/cerner/terra-framework/pull/1255/commits/9c1db0419e7aa76746431034d951a205600d5b6d

**Review by emilyrohrbough — APPROVED** [review]

File: packages/terra-application-navigation/src/ApplicationNavigation.jsx [hunk 6]
```diff
@@ -144,6 +144,8 @@ const ApplicationNavigation = ({
   const [drawerMenuIsOpen, setDrawerMenuIsOpen] = useState(false);
   const [popupMenuIsOpen, setPopupMenuIsOpen] = useState(false);
 
+  const closeMenuEventName = 'terra-application-navigation.dismiss-menu';
```
Comment (yuderekyu) [inline packages/terra-application-navigation/src/ApplicationNavigation.jsx:147]:
```suggestion
  const closeMenuEvent = 'terra-application-navigation.dismiss-menu';
```

File: packages/terra-application-navigation/src/ApplicationNavigation.jsx [hunk 7]
```diff
@@ -321,6 +323,19 @@ const ApplicationNavigation = ({
     }
   });
 
+  useEffect(() => {
+    const forceCloseMenu = () => {
+      setDrawerMenuIsOpen(false);
+      setPopupMenuIsOpen(false);
+    };
+
+    window.addEventListener(closeMenuEventName, forceCloseMenu);
```
Comment (yuderekyu) [inline packages/terra-application-navigation/src/ApplicationNavigation.jsx:332]:
```suggestion
    window.addEventListener(closeMenuEvent, forceCloseMenu);
```

File: packages/terra-application-navigation/src/ApplicationNavigation.jsx [hunk 8]
```diff
@@ -321,6 +323,19 @@ const ApplicationNavigation = ({
...
+
+    return () => {
+      window.removeEventListener(closeMenuEventName, forceCloseMenu);
```
Comment (yuderekyu) [inline packages/terra-application-navigation/src/ApplicationNavigation.jsx:335]:
```suggestion
      window.removeEventListener(closeMenuEvent, forceCloseMenu);
```

**Review by yuderekyu — APPROVED** [review]

**Pull request merged**